feat: HTTP Client Implementation with Java HTTP Client

sdelamo opened the pull request:
This PR adds a

- http-client-tck
- http-client-tck-netty
- http-client-tck-javanet
- http-client-javanet

The latter is an implementation of the Micronaut Framework Http Client using java.net.http.*

include

sdelamo committed:
wip: java.net.httpclient [ci skip]

This PR adds a

- http-client-tck
- http-client-tck-netty
- http-client-tck-javanet
- http-client-javanet

The latter is an implementation of the Micronaut Framework Http Client using java.net.http.*

include

timyates committed:
Get blocking delete working [skip ci]

timyates committed:
Async delete [skip ci]

timyates committed:
Simple body response

timyates committed:
Fix spotless

timyates committed:
Handle response body conversion [skip ci]

timyates committed:
Sealed classes [skip ci]

timyates committed:
Add a test for a declarative client, and get it passing [skip ci]

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/HttpRequestFactory.java:
```diff
@@ -35,7 +36,10 @@ private HttpRequestFactory() {
-        HttpRequest.BodyPublisher bodyPublisher = null;
+        HttpRequest.BodyPublisher bodyPublisher = request
+            .getBody()
+            .map(body -> HttpRequest.BodyPublishers.ofByteArray(ConversionService.SHARED.convertRequired(body, byte[].class)))
```
don't use the `SHARED` instance

timyates committed:
Use injected ConversionService where appropriate

timyates committed:
Merge branch '4.0.x' into javanet

timyates committed:
Checkstyle

timyates committed:
Suppress checkstyle for tck classes

timyates committed:
Test redirection

timyates committed:
Copyright [skip ci]

altro3 commented:
I wanted to say about one nuance in this client: there are restrictions for modifying headers. Some of them cannot be overridden. For me, this became a problem on one of the projects, when I needed to massively load http pages from different sites through a pool of proxy servers. Therefore, I eventually abandoned the use of this client.

I don't know if this will be a problem for you, but it's worth knowing about it.

timyates committed:
Add blocking redirect tests and fix url generation

sdelamo commented:
> I wanted to say about one nuance in this client: there are restrictions for modifying headers. Some of them cannot be overridden. For me, this became a problem on one of the projects, when I needed to massively load http pages from different sites through a pool of proxy servers. Therefore, I eventually abandoned the use of this client.

We don't play to replace the Micronaut HTTP Client based on netty but offer a HTTP Client variant based on built-in java HTTP Client.

sdelamo commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/AbstractJavanetHttpClient.java:
```diff
@@ -137,14 +151,41 @@ public void setMediaTypeCodecRegistry(MediaTypeCodecRegistry mediaTypeCodecRegis
+     * @return A URI that is prepended with the contextPath, if set
+     */
+    // TODO: Extract from DefaultHttpClient
+    protected URI prependContextPath(URI requestURI) {
```
Maybe it is worth having a `URIUtils` with a `prependUri` method?. I can see users using this when they want to redirect.

timyates committed:
Merge branch '4.0.x' into javanet

timyates committed:
Tidy up and add a declarative client for redirects [skip ci]

timyates committed:
Add a failing post body test [skip ci]

timyates committed:
Initial pass at POST bondy conversion [skip ci]

timyates committed:
Merge branch '4.0.x' into javanet

timyates committed:
Add server TCK with javanet client [skip ci]

timyates committed:
Handle status codes in return value and response [skip ci]

timyates committed:
Enable all status tests [skip ci]

timyates committed:
Error logging behind a check [skip ci]

timyates committed:
Cookie support [skip ci]

timyates committed:
Add Cookie test to client TCK [skip ci]

timyates committed:
Move helpers into the test classes to make things cleaner [skip ci]

timyates committed:
Add Experimental flag

timyates committed:
Merge branch '4.0.x' into javanet

timyates committed:
Checkstyle

timyates committed:
Sonar

timyates committed:
Expunge ConversionService.SHARED

timyates committed:
We have no sealed classes anymore to re-enable checkStyle and fix style errors

timyates committed:
2023

timyates committed:
Merge branch '4.0.x' into javanet

timyates committed:
Fixes for latest TCK

timyates commented:
Need to work on the SSL support

timyates committed:
Add initial docs for filling out

timyates committed:
Merge branch '4.0.x' into javanet

timyates committed:
BasicAuth test and initial SSL support.

timyates committed:
Checkstyle

timyates committed:
Http2 test

timyates committed:
Allow read-timeout and connect-timeout

timyates committed:
Allow http-proxy with the client

timyates committed:
Use the existing tck TestSuite

timyates committed:
Merge branch '4.0.x' into javanet

timyates committed:
Fix tests

timyates committed:
Remove sonar vulnerablities

timyates committed:
Work around sonar's optional checking

timyates committed:
Merge branch '4.0.x' into javanet

sdelamo commented on src/main/docs/guide/httpClient/javanetClient.adoc:
```diff
@@ -0,0 +1,3 @@
+As of Micronaut 4.0.0, you can use a java client based on java.net.HttpClient in place of the netty based default. To do so, replace the http-client dependency in your build with:
```
Using `Micronaut` as a noun is a violation of the brand guidelines

```suggestion
As of Micronaut Framework 4.0.0, you can use a java client based on java.net.HttpClient in place of the netty based default. To do so, replace the http-client dependency in your build with:
```

timyates committed:
Fix OctetTest

sdelamo commented on http-tck/src/main/java/io/micronaut/http/tck/ServerUnderTestProvider.java:
```diff
@@ -13,7 +13,7 @@
  * limitations under the License.
  */
-package io.micronaut.http.server.tck;
+package io.micronaut.http.tck;
```
I am concerned with this package change. It will break downstream implementations of the TCK

timyates committed:
Fix conversion for body assertion

timyates committed:
Fix body type for the other assert

sdelamo committed:
build: define junit.jupitr in libs.versions.toml (#8756)

sdelamo committed:
doc: reorder HTTP Clients implementations (#8754)

* link from the what's ne section
* moves  implementations to the top of the HTTP Client section
* Collapses the unsupported features in the same asciidoc
* Link to OpenJDK Java HTTP Client

sdelamo committed:
refactor: prefix tck implementations with tes-suite (#8755)

since these are modules which should not be published as artifacts and are just tests.

timyates committed:
Remove smells.  Failing here is a good thing as it's a test

timyates committed:
Add generic

timyates committed:
Update docs

sdelamo committed:
test: add tests for not follow redirects (#8759)

timyates committed:
Fix spotless

timyates committed:
Fix blockers

timyates committed:
Use a constant to make sonar happier

sdelamo committed:
refactor: extract CookieUtils (#8760)

sdelamo committed:
refactor: extract ContextPathUtils and HttpClientExceptionUtils (#8761)

- annotate ClientContextPathProvider with `@FunctionalInterface`
- anntate ServerContextPathProvider with `@FunctionalInterface`

sdelamo committed:
extract AbstractHttpClientFactory (#8765)

* change copyright year

* refactor: extract AbstractHttpClientFactory class

sdelamo commented:
@timyates several of the configuration settings apply only to the Netty implementation https://docs.micronaut.io/latest/guide/configurationreference.html#io.micronaut.http.client.DefaultHttpClientConfiguration

We should clarify which ones are only for the netty implementation at least in the Javadoc of DefaultHttpClientConfiguration.

sdelamo committed:
Extract response Adapter (#8766)

* change copyright year

* refactor: extract AbstractHttpClientFactory class

* refactor: extract response Adapter

sdelamo committed:
test:  micronaut.http.client.exception-on-error-status (#8767)

* test for exception-on-error-status config

* Fix for exception-on-error-status

---------

Co-authored-by: Tim Yates <tim.yates@gmail.com>

timyates commented:
Still need to do custom keystores

timyates committed:
Add read-timeout to requests

timyates committed:
Support loggerName config

timyates committed:
Spotless and CheckStyle appeasement

timyates committed:
Add apparently critical comment

timyates committed:
Use the correct plugin for the client-tck tests (#8789)

We we're adding them to the bom before

timyates committed:
Client tests use blocking and non-blocking client (#8790)

For netty, the blocking client is basically the non-blocking client with a Flux wrapper.

For javanet, this is not the case, and the blocking client makes use of the blocking methods in HttpClient.

To ensure both implementations are working, this PR adds a property to the server under test as to whether it should test with a blocking or non-blocking client.

timyates committed:
Add SSL support via SslBuilder implementation (#8779)

* Add SSL support via SslBuilder implementation

* Move hostname checking ssl tests to a new module

* Add why to the disableHostnameVerification detection

* Forgot an and

* Check the tests again, revoked works in the javanet client, so re-enable that test

* Add a pending feature for the non-functional ssl tests

sdelamo committed:
Merge branch '4.0.x' into javanet

timyates committed:
Mark config that isn't for javanet client as such (#8777)

* Mark config that isn't for javanet client as such

* Read-timeout is allowed

* Logger-name is allowed

* Switch logic to show where it's available

timyates committed:
Javanet client works with http proxies (#8781)

* Javanet client works with http proxies

* Fix for proxying with no loadbalancer

graemerocher commented on http-client-core/src/main/java/io/micronaut/http/client/AbstractHttpClientFactory.java:
```diff
@@ -0,0 +1,88 @@
+        this(null, ConversionService.SHARED);
+    }
+
+    @Creator
```
is `@Creator` needed here?

timyates replied:
Nope, nor was the empty constructor, so I removed both

graemerocher commented on http-client-core/src/main/java/io/micronaut/http/client/AbstractHttpClientFactory.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @NonNull
+    protected abstract T createHttpClient(@Nullable URI uri);
```
javadoc

graemerocher commented on http-client-core/src/main/java/io/micronaut/http/client/AbstractHttpClientFactory.java:
```diff
@@ -0,0 +1,88 @@
+    protected abstract T createHttpClient(@Nullable URI uri);
+
+    @NonNull
+    protected abstract T createHttpClient(@Nullable URI uri, @NonNull HttpClientConfiguration configuration);
```
javadoc

graemerocher commented on http-client-core/src/main/java/io/micronaut/http/client/exceptions/HttpClientExceptionUtils.java:
```diff
@@ -0,0 +1,51 @@
+ * @author Sergio del Amo
+ * @since 4.0.0
+ */
+public final class HttpClientExceptionUtils {
```
overkill to introduce a whole new class for one method. Just move the method to the parent abstract class

timyates replied:
As well as the JDK client, It's also used in 3 classes in the netty client, none of them share a parent class

yawkat replied:
can it be internal?

graemerocher commented on http-client-javanet/build.gradle.kts:
```diff
@@ -0,0 +1,24 @@
+plugins {
```
I feel like `javanet` is not a great name. Rename to `http-client-jdk`

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/AbstractJavanetHttpClient.java:
```diff
@@ -0,0 +1,232 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.client.javanet;
```
rename package from `javanet` to `jdk`

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/AbstractJavanetHttpClient.java:
```diff
@@ -0,0 +1,232 @@
+import static io.micronaut.http.client.exceptions.HttpClientExceptionUtils.populateServiceId;
+
+/*
+ * TODO: HttpClient defaults to netty ByteBuffer as a type for exchange which isn't best for here.
```
address this `TODO`

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/AbstractJavanetHttpClient.java:
```diff
@@ -0,0 +1,232 @@
+ * TODO: HttpClient defaults to netty ByteBuffer as a type for exchange which isn't best for here.
+ */
+@Experimental
+abstract class AbstractJavanetHttpClient {
```
rename to `AbstractJdkHttpClient`

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/AbstractJavanetHttpClient.java:
```diff
@@ -0,0 +1,232 @@
+
+    private final Logger log;
+
+    protected AbstractJavanetHttpClient(
```
javadoc

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/AbstractJavanetHttpClient.java:
```diff
@@ -0,0 +1,232 @@
+
+/*
+ * TODO: HttpClient defaults to netty ByteBuffer as a type for exchange which isn't best for here.
+ */
```
add `@since` and `@author`

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/AbstractJavanetHttpClient.java:
```diff
@@ -0,0 +1,232 @@
+        builder.sslParameters(sslParameters);
+    }
+
+    protected Object getLoadBalancerDiscriminator() {
+        return null;
+    }
+
+    public MediaTypeCodecRegistry getMediaTypeCodecRegistry() {
+        return mediaTypeCodecRegistry;
+    }
+
+    public void setMediaTypeCodecRegistry(MediaTypeCodecRegistry mediaTypeCodecRegistry) {
+        this.mediaTypeCodecRegistry = mediaTypeCodecRegistry;
+    }
```
javadoc public/protected methods

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/AbstractJavanetHttpClient.java:
```diff
@@ -0,0 +1,232 @@
+        builder.sslParameters(sslParameters);
+    }
+
+    protected Object getLoadBalancerDiscriminator() {
+        return null;
+    }
```
Since this is a new implementation, just remove this and pass the request as the discriminator. See https://github.com/micronaut-projects/micronaut-core/issues/8823

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/DefaultJavanetHttpClientRegistry.java:
```diff
@@ -0,0 +1,325 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Factory to create {@literal java.net.http.*} HTTP Clients.
```
add `@since`

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/DefaultJavanetHttpClientRegistry.java:
```diff
@@ -0,0 +1,325 @@
+
+    @Override
+    public void disposeClient(AnnotationMetadata annotationMetadata) {
+        //TODO
```
address this TODO

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/DefaultJavanetHttpClientRegistry.java:
```diff
@@ -0,0 +1,325 @@
+
+    @Override
+    public void close() throws Exception {
+        //TODO
```
Address this TODO and shutdown client

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/DefaultJavanetHttpClientRegistry.java:
```diff
@@ -0,0 +1,325 @@
+
+    @Override
+    public HttpClient getClient(HttpVersionSelection httpVersion, String clientId, String path) {
+        throw new UnsupportedOperationException("unimplemented");
```
implement this method

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/DefaultJavanetHttpClientRegistry.java:
```diff
@@ -0,0 +1,325 @@
+
+    @Override
+    public HttpClient resolveClient(InjectionPoint<?> injectionPoint, LoadBalancer loadBalancer, HttpClientConfiguration configuration, BeanContext beanContext) {
+        throw new UnsupportedOperationException("unimplemented");
```
implement this method

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/DefaultJavanetHttpClientRegistry.java:
```diff
@@ -0,0 +1,325 @@
+            configuration,
+            contextPath,
+            mediaTypeCodecRegistry,
+            beanContext.findBean(RequestBinderRegistry.class).orElseGet(() ->
+                new DefaultRequestBinderRegistry(conversionService)
+            ),
```
Replace these manual bean lookups with the `BeanProvider<RequestBinderRegistry>` injected into the constructor

timyates replied:
Done in https://github.com/micronaut-projects/micronaut-core/commit/e7a4fa8b2124bdc2af9aa0dd2ac6d78284f15e0b

There's no lazy orElse for BeanProvider, which means we will create a default each time, even if we don't need it

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/DefaultJavanetHttpClientRegistry.java:
```diff
@@ -0,0 +1,325 @@
+            ),
+            clientId,
+            conversionService,
+            beanContext.findBean(JavanetClientSslBuilder.class).orElseGet(() ->
```
Replace these manual bean lookups with the `BeanProvider<JavanetClientSslBuilder>` injected into the constructor

timyates replied:
Done in https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/e7a4fa8b2124bdc2af9aa0dd2ac6d78284f15e0b

There's no lazy orElse for BeanProvider, which means we will create a default each time, even if we don't need it

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/DefaultJavanetHttpClientRegistry.java:
```diff
@@ -0,0 +1,325 @@
+     * Client key.
+     */
+    @Internal
+    private static final class ClientKey {
```
can be a `record`

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/HttpCookieUtils.java:
```diff
@@ -0,0 +1,53 @@
+ * @author Sergio del Amo
+ * @since 4.0.0
+ */
+public final class HttpCookieUtils {
```
again overkill a whole new class for 1 static method

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/HttpResponseAdapter.java:
```diff
@@ -0,0 +1,107 @@
+
+    @Override
+    public String reason() {
+        throw new UnsupportedOperationException("Not implemented yet");
```
address this

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/HttpResponseAdapter.java:
```diff
@@ -0,0 +1,107 @@
+
+    @Override
+    public MutableConvertibleValues<Object> getAttributes() {
+        return null;
```
support attributes

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/JavanetBlockingHttpClient.java:
```diff
@@ -0,0 +1,91 @@
+                throw HttpClientExceptionUtils.populateServiceId(new HttpClientResponseException(HttpStatus.valueOf(httpResponse.statusCode()).getReason(), response(httpResponse, bodyType)), clientId, configuration);
+            }
+            return response(httpResponse, bodyType);
+        } catch (IOException | InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new HttpClientException("error sending request", e);
+        }
```
you shouldn't interrupt the thread on `IOException` only. Also include the message from the original exception in the message

timyates replied:
Fixed (I believe) by https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/a53d3a81b06a96d8ee2dc2a76d7d0f1f8f117a79

graemerocher commented on http-client-javanet/src/main/java/io/micronaut/http/client/javanet/JavanetHttpClient.java:
```diff
@@ -0,0 +1,150 @@
+                }
+                return client.sendAsync(httpRequest, java.net.http.HttpResponse.BodyHandlers.ofByteArray());
+            })
+            .flatMap(Mono::fromFuture)
```
this is not idea because `fromFuture` will block waiting for the future to complete. You want to use `fromCompletionStage` instead which can register a listener. Also to support the different `exchangeStream` use cases of the HTTP client you will need to use `HttpResponse.BodyHandlers.ofPublisher()` and include the `Flow.Publisher` adapters.

timyates replied:
Good knowledge on completionStage vs Future.  Fixed that in https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/72431a66a6a3b6c2d22a9c68235f5f045853b1d4

I'm not sure of the second part?  How do I "_…include the `Flow.Publisher` adapters_"?

timyates replied:
Relevant code (since refactoring) [is here](https://github.com/micronaut-projects/micronaut-core/blob/332cfd3278089dd5d14a245fe91609f0cae82f6e/http-client-jdk/src/main/java/io/micronaut/http/client/jdk/JdkHttpClient.java#L131-L152)

graemerocher reviewed (changes requested):
There seems to be a lot missing still (`StreamingHttpClient`, `WebSocketClient` etc.) but seems like a good start.

timyates committed:
javanet -> jdk

timyates committed:
Javadoc

timyates committed:
Rename packages

timyates committed:
Pull cookie conversion into abstract class

timyates committed:
Use request as discriminator

timyates committed:
Switch to record

timyates committed:
Rename from javanet

timyates committed:
Implement close and disposeClient

timyates committed:
Tag multipart test unsupported by JDK client

timyates committed:
Implement getClient

timyates committed:
Inject bean providers instead of using a lookup

timyates committed:
Support reason and attributes

timyates committed:
Fix exception handling

timyates committed:
Merge branch '4.0.x' into javanet

timyates committed:
Fix merge for filter tests

timyates committed:
Ignore ClientFilter tests while I work on them

timyates commented:
Raised https://github.com/micronaut-projects/micronaut-core/issues/8844 to handle ClientFilters

timyates committed:
Switched to fromCompletionStage as per review

timyates committed:
Remove unused constructor

timyates committed:
Javadoc

timyates committed:
Whitespace

yawkat commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/AbstractJdkHttpClient.java:
```diff
@@ -0,0 +1,277 @@
+        configuration.getConnectTimeout().ifPresent(builder::connectTimeout);
+
+        builder
+            .version(httpVersion != null && httpVersion.isAlpn() ? HttpClient.Version.HTTP_2 : HttpClient.Version.HTTP_1_1)
```
there should be more checks here

- h2c is not supported by the jdk (check plaintextMode)
- the user could enable alpn without h2, in which case http1 should be used

the first may be a bit annoying. if you set http2 in the config using the old http version field, it will enable h2c, so this would error. but i think it's correct to error here.

sdelamo replied:
Based on @yawkat feedback probably best to move it to a protected method someone may override if we miss anything.

yawkat replied:
i dont think it's necessary, it is already possible to set the right config to get the jdk client version you want. i just think the heuristic should be a bit different to closer match the netty client

timyates replied:
Like this you mean? https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/9174a0026a2f5b2ecb375f4ced4da30accf496e3

I'm a bit lost in the many, many, many ways we have for configuring this...

yawkat replied:
ok first you have to extract the version selection using this method: https://github.com/micronaut-projects/micronaut-core/blob/d56c5abdac14bf1e4e9f3993c6e1febc3d6a0896/http-client-core/src/main/java/io/micronaut/http/client/HttpVersionSelection.java#L96

this handles the different old/new config properties.

then you have a plaintextMode, and alpn settings (for tls). for plaintextMode, only http1 is supported, so please verify this. for alpn, you can configure the client with http2 if alpn is enabled *and* the alpn contains ALPN_HTTP_2.

timyates replied:
Like this? https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/756045fa30b246615f8ed23b06532dc742745a02

yawkat commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/AbstractJdkHttpClient.java:
```diff
@@ -0,0 +1,277 @@
+
+    private void configureSsl(HttpClient.Builder builder, ClientSslConfiguration clientSslConfiguration) {
+        sslBuilder.build(clientSslConfiguration).ifPresent(builder::sslContext);
+        if (System.getProperty("jdk.internal.httpclient.disableHostnameVerification") != null && log.isWarnEnabled()) {
```
should be moved to constructor. this method is only called when there is an ssl config

yawkat commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/DefaultJdkHttpClientRegistry.java:
```diff
@@ -0,0 +1,328 @@
+@BootstrapContextCompatible
+@Internal
+@Experimental
+public class DefaultJdkHttpClientRegistry implements AutoCloseable, HttpClientRegistry<HttpClient> {
```
make final

timyates committed:
Make Default client registry final as per review

timyates committed:
Move warning to constructor so it always shows

timyates committed:
Throw error if H2C is configured as it's not supported

timyates committed:
Factory needs a no-arg constructor for HttpClient.create calls

timyates committed:
Fix logger name from config

yawkat commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/DefaultJdkHttpClientRegistry.java:
```diff
@@ -0,0 +1,328 @@
+    }
+
+    private HttpClient getClient(ClientKey key) {
+        return clients.computeIfAbsent(key, clientKey -> {
```
do we do this similarly in the netty client?

computeIfAbsent can lead to deadlocks if it is called recursively, so it seems a bit dangerous to have such a big lambda. but maybe it's fine.

timyates replied:
Yes

https://github.com/micronaut-projects/micronaut-core/blob/4ec18d886a0c705b5f2ea91516dfaee95644cbcc/http-client/src/main/java/io/micronaut/http/client/netty/DefaultNettyHttpClientRegistry.java#L310

It's 9 lines longer for Netty

yawkat replied:
ok, then just ignore it :D

yawkat commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/HttpResponseAdapter.java:
```diff
@@ -0,0 +1,110 @@
+ * @since 4.0.0
+ * @param <O> Body Type
+ */
+public class HttpResponseAdapter<O> implements HttpResponse<O> {
```
can this be internal

timyates replied:
Yup https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/fe01ab2f5c046348a1db5a72003cc02180a4527d

yawkat commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/JdkClientSslBuilder.java:
```diff
@@ -0,0 +1,120 @@
+        TrustManagerFactory trustManagerFactory = getTrustManagerFactory(ssl);
+        KeyManagerFactory keyManagerFactory = getKeyManagerFactory(ssl);
+        try {
+            SSLContext tls = SSLContext.getInstance(ssl.getProtocol().orElse("TLS"));
```
maybe if HttpVersionSelection.http2CipherSuites is set, we can limit the cipher suites to those specified by http2 here. but maybe the jdk client already does this internally.

yawkat commented on http-client-jdk/src/test/groovy/io/micronaut/http/client/jdk/ClientProxySpec.groovy:
```diff
@@ -0,0 +1,127 @@
+        }
+    }
+
+    private static void resetSystemProperty(String property, String value) {
```
isnt there a spock feature for this

timyates replied:
Good catch https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/32e5fa22aec2b6565dc0248a2d0989f0f1deda4e

yawkat commented on http-client/src/test/resources/logback.xml:
```diff
@@ -11,7 +11,5 @@
-
-<!--    <logger name="io.micronaut.http.client" level="trace" />-->
-
+    <logger name="io.micronaut.http.client" level="DEBUG" />
```
revert

timyates replied:
https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/f1b8f378413fd57c32fd549231a87b2c3adf1594

yawkat commented on test-suite-http-client-jdk-ssl/src/test/resources/logback.xml:
```diff
@@ -0,0 +1,16 @@
+        <appender-ref ref="STDOUT" />
+    </root>
+
+    <logger name="io.micronaut.http.client" level="TRACE" />
```
disable

timyates replied:
https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/8b2fe74bfb2c1c883d0215d8550b9d07c99417f8

yawkat commented on test-suite-http-client-tck-jdk/src/test/resources/logback.xml:
```diff
@@ -0,0 +1,15 @@
+    <root level="info">
+        <appender-ref ref="STDOUT" />
+    </root>
+    <logger name="io.micronaut.http.client" level="TRACE" />
```
disable

yawkat commented on test-suite-http-client-tck-netty/src/test/resources/logback.xml:
```diff
@@ -0,0 +1,15 @@
+    <root level="info">
+        <appender-ref ref="STDOUT" />
+    </root>
+    <logger name="io.micronaut.http.client" level="trace" />
```
remove

yawkat commented on test-suite-http-server-tck-jdk/build.gradle.kts:
```diff
@@ -0,0 +1,15 @@
+}
+
+tasks.withType(Test::class) {
+    systemProperty("jdk.httpclient.HttpClient.log", "all")
```
remove

yawkat commented on test-suite-http-server-tck-jdk/src/test/resources/logback.xml:
```diff
@@ -0,0 +1,11 @@
+    <root level="info">
+        <appender-ref ref="STDOUT" />
+    </root>
+    <logger name="io.micronaut.http.client" level="TRACE" />
```
remove

yawkat commented on test-suite-http-server-tck-netty/src/test/resources/logback.xml:
```diff
@@ -7,4 +7,5 @@
     <root level="info">
         <appender-ref ref="STDOUT" />
     </root>
+    <logger name="io.micronaut.http.client" level="TRACE" />
```
remove

timyates committed:
Mark as internal

timyates committed:
Revert logging change

timyates committed:
Turn down logging

timyates committed:
Switch to spock feature for system props

timyates commented:
@graemerocher Do you want to take one last pass prior to merging?

timyates committed:
Correctly select version from config

timyates commented:
I think https://github.com/micronaut-projects/micronaut-core/pull/8441#discussion_r1118828543 is outstanding, but I'm unsure what to do

timyates committed:
Make HttpClientExceptionUtils internal

graemerocher commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/AbstractJdkHttpClient.java:
```diff
@@ -0,0 +1,294 @@
+     * @param <I>      The body type
+     * @return A JDK request object
+     */
+    protected <I> Flux<HttpRequest> mapToHttpRequest(io.micronaut.http.HttpRequest<I> request, Argument<?> bodyType) {
```
Use `Mono` here since there is only one result

graemerocher commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/AbstractJdkHttpClient.java:
```diff
@@ -0,0 +1,294 @@
+            });
+    }
+
+    private Flux<URI> resolveRequestUri(io.micronaut.http.HttpRequest<?> request) {
```
Use `Mono` here since there is only one result

graemerocher commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/AbstractJdkHttpClient.java:
```diff
@@ -0,0 +1,294 @@
+        return resolveURI(request);
+    }
+
+    private <I> Flux<URI> resolveURI(io.micronaut.http.HttpRequest<I> request) {
```
Use `Mono` here since there is only one result

graemerocher commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/HttpResponseAdapter.java:
```diff
@@ -0,0 +1,114 @@
+            } else if (type.getType() == byte[].class) {
+                return Optional.of(bytes);
+            } else {
+                Optional<MediaTypeCodec> foundCodec = mediaTypeCodecRegistry.findCodec(contentType);
```
rewrite using `flatMap`

timyates replied:
Like so?  I couldn't use `flatMap` as we have the catch-all method at the end of the function https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/9fbb719240df57c9cc8f5f0b040aed0de8927d84

graemerocher commented on src/main/docs/guide/httpClient/httpClientImplementations/javanetClient.adoc:
```diff
@@ -0,0 +1,14 @@
+To use an implementation based on https://openjdk.org/groups/net/httpclient/intro.html[Java HTTP Client], add the following dependency to your build:
+
+dependency:micronaut-http-client-javanet[]
```
correct this to reference the renamed module

graemerocher commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/DefaultJdkHttpClientRegistry.java:
```diff
@@ -0,0 +1,328 @@
+     */
+    @Bean
+    @BootstrapContextCompatible
+    @Primary
```
what happens if both clients on the classpath? Perhaps `@Order` should be used here to define priority

timyates replied:
👍

graemerocher commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/JdkHttpClient.java:
```diff
@@ -0,0 +1,158 @@
+ */
+@Internal
+@Experimental
+public class JdkHttpClient extends AbstractJdkHttpClient implements HttpClient {
```
rename this to `DefaultJdkHttpClient` and add an interface called `JdkHttpClient` that would let me inject if I have both clients on the classpath

timyates replied:
I think this fixes it

https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/69da8ad28ab427d143124d7a6c96fb1d5768752c

There's some nastiness around extracting cookies from the Netty client request, but I added tests to the test-suites (and pulled both client in to check JdkHttpClient injection)

timyates committed:
Switch to Mono for single things

timyates committed:
Use Optional.map

timyates committed:
Rename docs to match the code

timyates committed:
Allow multiple clients to be on the classpath

timyates committed:
Merge branch '4.0.x' into javanet

timyates committed:
Add a couple of tests around attributes

timyates committed:
Another attribute test
